### ISSUE TITLE
Python 3 compatibility

### ISSUE THREAD

**Issue by niko** [signalfuse/maestro-ng#27]
In python 3 I get ```NameError: name 'reduce' is not defined``` when I try to run Maestro:

```
 % python -m maestro -f Maestrofile start webapp                                                                                                        
Traceback (most recent call last):
  File "/usr/lib/python3.4/runpy.py", line 171, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.4/runpy.py", line 86, in _run_code
    exec(code, run_globals)
  File "/home/niko/.local/lib/python3.4/site-packages/maestro/__main__.py", line 96, in <module>
    sys.exit(main(sys.argv[1:]))
  File "/home/niko/.local/lib/python3.4/site-packages/maestro/__main__.py", line 88, in main
    getattr(c, options.command)(**vars(options))
  File "/home/niko/.local/lib/python3.4/site-packages/maestro/maestro.py", line 209, in start
    if not only else self._to_containers(things)
  File "/home/niko/.local/lib/python3.4/site-packages/maestro/maestro.py", line 167, in _ordered_containers
    sorted(self._gather_dependencies(self._to_containers(things),
  File "/home/niko/.local/lib/python3.4/site-packages/maestro/maestro.py", line 156, in _to_containers
    return reduce(lambda x, y: x+y, map(parse_thing, things), [])
NameError: name 'reduce' is not defined
```

The error occurs here: https://github.com/signalfuse/maestro-ng/blob/dc6e2030295586f9072de8eaf967ebadafea0622/maestro/maestro.py#L156

```reduced``` was moved to functools in python 3.

Just a guess (as I know close to no python), but I think the same applies here:

https://github.com/signalfuse/maestro-ng/blob/dc6e2030295586f9072de8eaf967ebadafea0622/maestro/maestro.py#L140

and here:

https://github.com/signalfuse/maestro-ng/blob/ff32775c96f5e5625a720c4cecc4fab4da9ae662/maestro/plays.py#L217

All the best, Niko.

**Comment by elvetemedve** [comment]
In Python 3.4.3 I got an exception for `python -m maestro start`, while the same in 2.7.9 run without any error.
Version: maestro-0.2.6-dev
Here is the stack trace from the exception:
```
Traceback (most recent call last):

  File "/home/geza/.local/lib/python3.4/site-packages/maestro/maestro.py", line 249, in _audit_play
    play.run()
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/plays/__init__.py", line 146, in run
    self._end()
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/plays/__init__.py", line 138, in _end
    exceptions.raise_with_tb(self._error)
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/exceptions.py", line 19, in raise_with_tb
    raise info[1].with_traceback(info[2])
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/plays/__init__.py", line 95, in act
    task.run()
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/plays/tasks.py", line 130, in run
    result = self._create_and_start_container()
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/plays/tasks.py", line 201, in _create_and_start_container
    command=self.container.command)
  File "/home/geza/.local/lib/python3.4/site-packages/docker/client.py", line 463, in create_container
    return self.create_container_from_config(config, name)
  File "/home/geza/.local/lib/python3.4/site-packages/docker/client.py", line 470, in create_container_from_config
    res = self._post_json(u, data=config, params=params)
  File "/home/geza/.local/lib/python3.4/site-packages/docker/client.py", line 143, in _post_json
    return self._post(url, data=json.dumps(data2), **kwargs)
  File "/usr/lib/python3.4/json/__init__.py", line 230, in dumps
    return _default_encoder.encode(obj)
  File "/usr/lib/python3.4/json/encoder.py", line 192, in encode
    chunks = self.iterencode(o, _one_shot=True)
  File "/usr/lib/python3.4/json/encoder.py", line 250, in iterencode
    return _iterencode(o, 0)
  File "/usr/lib/python3.4/json/encoder.py", line 173, in default
    raise TypeError(repr(o) + " is not JSON serializable")
TypeError: <map object at 0x7f6b30222908> is not JSON serializable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/__main__.py", line 204, in execute
    getattr(c, options.command)(**vars(options))
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/maestro.py", line 331, in start
    ignore_dependencies, concurrency, reuse))
  File "/home/geza/.local/lib/python3.4/site-packages/maestro/maestro.py", line 261, in _audit_play
    exceptions.raise_with_tb()
TypeError: raise_with_tb() missing 1 required positional argument: 'info'
```

**Comment by mpetazzoni** [comment]
Thanks! The `raise_with_tb()` error is definitely something in Maestro I need to fix. The original error though, that might be in `docker-py`. I'll take a look and fix the bug there is one/if necessary.

**Comment by binary1230** [comment]
argh. would be cool if the docs said "don't use python 3", I have been fighting this all day trying to get a setup going and hitting this and some other issues

**Comment by mpetazzoni** [comment]
@binary1230 Do you have some stacktraces of errors with Python3? I'd love to fix those and make sure Maestro works with Python3 correctly! I guess I should spend some time making sure it all works with Python3 and write more unit tests; the ones I have right now all pass, but they also don't have very broad coverage.

**Comment by binary1230** [comment]
oh dang, I actually deleted the VM I was working on today.

though, the only thing I ran into that actually caused any problems was the stacktrace in this issue (it's how I found this issue, googling it).  I got stuck here and didn't get it to go further. (I've since deleted the VM and did it all with python2, that's working great)

a bunch of my other stuff was caused just by fighting my ubuntu installation with python3 vs 2 and installing a newer version of pip, it wasn't maestro-ng's fault.

everything else is working great so far though, nice work on this project, I really like the design of everything so far.
